### DISCUSSION TITLE
Help with some of the functions / example code?

### DISCUSSION THREAD

**Discussion by evanmang** [iluvcapra/py-ptsl#47] [Q&A] [answered]
First off THANK YOU. I've been able to do SO MUCH with this tool already that would have never been possible with Applescript and the methods I was previously using.

My python is mediocre _at best_, and I'm having trouble with some of the functions that take multiple parameters.

For example, i'm able to run `create_memory_location` and it creates a memory location at the current timeline with the default settings, but I can't figure out the syntax of how I would pass the other parameters (`name, start_time`) to the function. I recognize that this is likely due to my ignorance as to how these data structures work in python in general, but a simple example of how that would look would help me (and hopefully others) a ton!

**Comment by iluvcapra** [comment]
Hi Evan-

Thanks for checking out the library!

A lot of python code has inline documentation, and a lot of more modern python code has type annotations for things like function parameters. If you look at the `create_memory_location` definition:

https://github.com/iluvcapra/py-ptsl/blob/master/ptsl/engine.py#L359

`start_time` and `end_time` are `Optional[str]`, which means you have to pass these as strings (you just pass a time in the current ruler format, like `"1:00:00:00"`).

**Reply by evanmang** [reply]
Thanks for the reply! I saw the documentation in the engine.py file, but it seems like i'm just misunderstanding the correct syntax to pass those parameters.

for example:

`    engine.create_memory_location(start_time="01:00:00:00", memory_number=100, name="test")
`
Does give me a new memory location with memory location number 100, but with a default name, and at the current timecode position.

**Reply by StephenMachete** [reply, marked as answer]
Hi Evan,
i can't reproduce the Problem. Do you get any error? Something like:
"ptsl.errors.CommandError: ErrType 100: PT_UnknownError (zoom_settings)"
Do you have the latest Version of py-ptsl installed? (401)
What Version of PT do you use?

Your code is fine, the syntax is correct. But you should get the error i mentioned above, because of an unresolved bug. The error should also be the Reason the Marker isnt placed correctly. If i run your code, i get the error and a marker at current timecode position with number 100 and name test.

To fix your problem temporarily (please be aware that this is only a workaround and comes with some other issues for some features of the library):
You can go to the client.py in the library and add the "always_print_fields_with_no_presence=True" attribute to the "MessageToJson" in the "_prepare_operation_request_json" function:


It should look like this:
```
    def _prepare_operation_request_json(self, operation):
        """
        Convert the request body into a JSON string (or an empty string if
        there is no request body.
        """
        if operation.request is None:
            request_body_json = ""
        else:
            request_body_json = \
                json_format.MessageToJson(operation.request,
                                          always_print_fields_with_no_presence=True,
                                          preserving_proto_field_name=True)
```

This should fix your problem.

Best
Steve

**Reply by iluvcapra** [reply]
Yeah this did it, this `always_print_fields_with_no_presence` has to be set True, I'll make this change in the source.

**Comment by evanmang** [comment]
Steve,

Thanks for your reply. I'm running the latest py-ptsl, and Pro Tools 2024.6. I'm indeed getting the same error!

`ptsl.errors.CommandError: ErrType 100: PT_UnknownError (zoom_settings)`

I added that line to the cliient.py file and the function now works,

-Evan

**Comment by iluvcapra** [comment]
I've made a small change in client.py that fixes this, it's presently on master can you give it a try and let me know if it works for you?
